types/uni-app/lib/uni.d.ts

```diff
@@ -5477,13 +5475,24 @@ interface CanvasContext {
     /**
      * 对指定的图像创建模式的方法，可在指定的方向上重复元图像
      */
-    createPattern(image?: string, repetition?: 'repeat' | 'repeat-x' | 'repeat-y' | 'no-repeat'): void;
+    createPattern(image?: string, repetition?: 'repeat' | 'repeat-x' | 'repeat-y' | 'no-repeat'): CanvasPattern | null;
     /**
      * 使用矩阵重新设置（覆盖）当前变换的方法
      */
     setTransform(scaleX?: number, skewX?: number, skewY?: number, scaleY?: number, translateX?: number, translateY?: number): void;
 }
 
+interface CanvasGradient {
+    /**
+     * 创建一个颜色的渐变点
+     */
+    addColorStop(stop?: number, color?: string): void;
+}
+
+// tslint:disable-next-line no-empty-interface
```

Comment:
What's the purpose of this empty interface?

Reply:
[Doc](https://uniapp.dcloud.io/api/canvas/CanvasContext?id=canvascontextcreatepattern)

Below is the example code I copy from the doc of uni-app.
```js
const ctx = uni.createCanvasContext('myCanvas')
const pattern = ctx.createPattern('/path/to/image', 'repeat-x')
ctx.fillStyle = pattern
ctx.fillRect(0, 0, 300, 150)
ctx.draw()
```
We can get the `CanvasPattern` object by calling `uni.createCanvasContext`, and assign it to `ctx.fillStyle`. The `CanvasPattern` has no public props or methods and could only obtian from calling `uni.createCanvasContext`. We could not assign an any object (`{}`) to `ctx.fillStyle`.

It's somewhat like the `WebGLObject` definition in "lib.dom.d.ts"
```ts
interface WebGLObject {
}
```
No public props or methods, but it is special.

Please forgive my pool english.

Reply:
I think an easy approximation to the behavior you want would be to use `object` (e.g. `type CanvasPattern = object`).  A fancier version might be a class with private field (e.g. `image` and `repetition`).

Reply:
While I'm pretty sure that type aliases do appear in intellisense, I assume you've tested out the experience and confirmed that the empty interface works better in some scenarios.

I'm not sure the private field approach is a "hack" since the underlying library does appear to [declare ](https://github.com/dcloudio/uni-app/blob/1de259ec9eebd1d41951d0b8ffc985b6c2e9dd4f/src/core/service/api/context/canvas.js#L242)the type as a class with two private fields.

Note that an empty interface provides little type safety beyond using `any` - it mostly makes tool tips more readable.

Reply:
> While I'm pretty sure that type aliases do appear in intellisense, I assume you've tested out the experience and confirmed that the empty interface works better in some scenarios.

Type aliases do appear in intellisense while they denote a "not so complicate" type.

See following cases:
```typescript
type TA = object;
let a: TA; // when hover, shows 'let a: object'

type TB = () => object;
let b: TB; // when hover, shows 'let b: TB'

let c: TA | TB; // when hover, shows 'let c: object | TB'
```

> I'm not sure the private field approach is a "hack" since the underlying library does appear to declare the type as a class with two private fields.

[Uniapp](https://uniapp.dcloud.io/README) is a cross-platform framework. Code built with uniapp runs not only on browser but also on several different miniprogram platforms. While the underlying library which is just the runtime for browser uses a class as implementation, it may be different in other runtimes which are not open source.

Thanks for your approval. 😀